patches/server/0965-Make-banners-modifiable-with-both-BannerMeta-and-Blo.patch

```diff
@@ -0,0 +1,274 @@
+--- a/src/main/java/org/bukkit/craftbukkit/inventory/CraftMetaBanner.java
++++ b/src/main/java/org/bukkit/craftbukkit/inventory/CraftMetaBanner.java
+@@ -18,7 +18,7 @@ import org.bukkit.craftbukkit.util.CraftMagicNumbers;
+ import org.bukkit.inventory.meta.BannerMeta;
+ 
+ @DelegateDeserialization(CraftMetaItem.SerializableMeta.class)
+-public class CraftMetaBanner extends CraftMetaItem implements BannerMeta {
++public class CraftMetaBanner extends CraftMetaBlockState implements BannerMeta { // Paper - Make banners also modifiable with BlockStateMeta
+ 
+     private static final Set<Material> BANNER_MATERIALS = Sets.newHashSet(
+             Material.BLACK_BANNER,
+@@ -60,13 +60,21 @@ public class CraftMetaBanner extends CraftMetaItem implements BannerMeta {
+     static final ItemMetaKey COLOR = new ItemMetaKey("Color", "color");
+     static final ItemMetaKey PATTERN = new ItemMetaKey("Pattern", "pattern");
+ 
++    // Paper start - Add banner materials to CraftMetaBlockState's material list
++    static {
++        CraftMetaBlockState.BLOCK_STATE_MATERIALS.addAll(CraftMetaBanner.BANNER_MATERIALS);
++    }
++    // Paper end - Add banner materials to CraftMetaBlockState's material list
++
+     private DyeColor base;
+     private List<Pattern> patterns = new ArrayList<Pattern>();
+ 
+-    CraftMetaBanner(CraftMetaItem meta) {
+-        super(meta);
++    // Paper start - Create new constructor to properly handle different banners
++    CraftMetaBanner(CraftMetaItem meta, Material material) {
++        super(meta, material);
+ 
+-        if (!(meta instanceof CraftMetaBanner)) {
++        if (true || !(meta instanceof CraftMetaBanner)) {
++            // Paper end - Create new constructor to properly handle different banners
+             return;
+         }
+ 
+@@ -75,10 +83,12 @@ public class CraftMetaBanner extends CraftMetaItem implements BannerMeta {
+         this.patterns = new ArrayList<Pattern>(banner.patterns);
+     }
+ 
+-    CraftMetaBanner(CompoundTag tag) {
+-        super(tag);
++    // Paper start - Create new constructor to properly handle different banners
++    CraftMetaBanner(CompoundTag tag, Material material) {
++        super(tag, material);
+ 
+-        if (!tag.contains("BlockEntityTag")) {
++        if (true || !tag.contains("BlockEntityTag")) {
++            // Paper end - Create new constructor to properly handle different banners
+             return;
+         }
+ 
+@@ -102,10 +112,19 @@ public class CraftMetaBanner extends CraftMetaItem implements BannerMeta {
+ 
+     CraftMetaBanner(Map<String, Object> map) {
+         super(map);
++        // Paper start - Correctly deserialize old BannerMeta to new BlockStateMeta
++        if (map.containsKey("blockMaterial")) {
++            return;
++        }
++        this.material = Material.WHITE_BANNER;
++        if (map.containsKey("internal")) {
++            return;
++        }
++        // Paper end - Correctly deserialize old BannerMeta to new BlockStateMeta
```

Comment:
I think a better solution to this is to modify the map going to CraftMetaBlockState by inserting the `blockMaterial` if it doesn't exist. Something like
```java
    private static Map<String, Object> insertDefaultMaterial(Map<String, Object> map) {
        if (map.containsKey("blockMaterial")) {
            return map;
        } else {
            final Map<String, Object> newMap = new HashMap<>(map);
            newMap.put("blockMaterial", "minecraft:white_banner");
            return newMap;
        }
    }
```
and then wrapping the map in the super ctor in that method. That way you don't have to remove the `final` from the Material field in the CraftMetaBlockState.

Reply:
Thought of this but struggled to choose between this and the mindiff, and ended up using the mindiff one. Switching to this!